src/executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+    log::info!(
+        "initialized compute-heavy future executor strategy - {:#?}",
+        global_sync_strategy()
+    );
+
+    Ok(())
+}
+
+/// Get the currently initialized sync strategy,
+/// or the default strategy for the current feature in case no strategy has been loaded.
+///
+/// See [`SyncExecutorBuilder`] for details on strategies.
+///
+/// # Examples
+///
+/// ```
+/// use compute_heavy_future_executor::{
+///     global_sync_strategy,
+///     global_sync_strategy_builder,
+///     GlobalStrategy,
+///     ExecutorStrategy
+/// };
+///
+/// # fn run() {
+///
+/// #[cfg(feature = "tokio")]
+/// assert_eq!(global_sync_strategy(), GlobalStrategy::Default(ExecutorStrategy::SpawnBlocking));
+///
+/// #[cfg(not(feature = "tokio"))]
+/// assert_eq!(global_sync_strategy(), GlobalStrategy::Default(ExecutorStrategy::CurrentContext));
+///
+/// global_sync_strategy_builder()
+///         .initialize_current_context()
+///         .unwrap();
+///
+/// assert_eq!(global_sync_strategy(), GlobalStrategy::Initialized(ExecutorStrategy::CurrentContext));
+///
+/// # }
+/// ```
+pub fn global_sync_strategy() -> GlobalStrategy {
+    match COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY.get() {
+        Some(strategy) => GlobalStrategy::Initialized(strategy.into()),
+        None => GlobalStrategy::Default(<&SyncExecutor>::default().into()),
+    }
+}
+
+pub(crate) fn get_global_sync_executor() -> &'static SyncExecutor {
+    COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY
+        .get()
+        .unwrap_or_else(|| <&SyncExecutor>::default())
+}
+
+/// The stored strategy used to spawn compute-heavy futures.
+static COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY: OnceLock<SyncExecutor> = OnceLock::new();
+
+/// The fallback strategy used in case no strategy is explicitly set
+static DEFAULT_COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY: OnceLock<SyncExecutor> = OnceLock::new();
+
+#[non_exhaustive]
+pub(crate) enum SyncExecutor {
+    /// A non-op strategy that runs the function in the current context
+    CurrentContext(current_context::CurrentContextExecutor),
+    /// User-provided closure
+    CustomExecutor(custom_executor::CustomExecutor),
+    /// tokio task::spawn_blocking
+    #[cfg(feature = "tokio")]
+    SpawnBlocking(spawn_blocking::SpawnBlockingExecutor),
+}
+
+impl Default for &SyncExecutor {
+    fn default() -> Self {
+        DEFAULT_COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY.get_or_init(|| {
+            let core_count = num_cpus::get();
+
+            #[cfg(feature = "tokio")]
+            {
+                log::info!("Defaulting to SpawnBlocking strategy for compute-heavy future executor \
+                with max concurrency of {core_count} until a strategy is initialized");
+
+                SyncExecutor::SpawnBlocking(spawn_blocking::SpawnBlockingExecutor::new(Some(core_count)))
+            }
+
+            #[cfg(not(feature = "tokio"))]
+            {
+                log::warn!("Defaulting to CurrentContext (non-op) strategy for compute-heavy future executor \
```

Comment:
i was thinking about this more, and actually we probably want FEWER than the cores by default for current context, so that we make sure we don't fully starve the tokio executor. core - 2 or something. though for current thread tokio runtime, that breaks anyway.